File: 2020/day3/day3.go
package main

import (
	"fmt"
	"github.com/kamillo/aoc/utils"
)

type slope struct {
	right int
	down  int
}

func main() {
	lines := utils.GetLines("input.txt")

	fmt.Println("Part 1: ", countTrees(lines, slope{3, 1}))

	// Part 2
	slopes := []slope{
		{1, 1},
		{3, 1},
		{5, 1},
		{7, 1},
		{1, 2},
	}

	treesProduct := 1
	for _, slope := range slopes {
		treesProduct *= countTrees(lines, slope)
	}

	fmt.Println("Part 2: ", treesProduct)
}

func countTrees(grid []string, slope slope) (treesCount int) {
	x, y := 0, 0
	width := len(grid[0])

	for y < len(grid) {
		if grid[y][x%width] == '#' {
			treesCount++
		}
		y += slope.down
		x += slope.right
	}

	return
}

File: nim/2020/day3/day3.nim
import utils/fileutils

type 
    Slope = object
        right: int
        down: int

proc countTrees(grid: seq[string], slope: Slope): int =
    var
        x = 0
        y = 0
        width = grid[0].len

    while y < grid.len:
        if grid[y][x mod width] == '#':
            result += 1
        y += slope.down
        x += slope.right

var 
    lines = getLines("input.txt")
    slopes: seq[Slope] = @[
        Slope(right: 1, down: 1),
        Slope(right: 3, down: 1),
        Slope(right: 5, down: 1),
        Slope(right: 7, down: 1),
        Slope(right: 1, down: 2),
    ]
    treesProduct = 1

echo("Part 1: ", countTrees(lines, Slope(right: 3, down: 1)))

for slope in slopes:
    treesProduct *= countTrees(lines, slope)

echo("Part 2: ", treesProduct)